UI Component 중 DropDown 구현

seonyoung42 opened the pull request:
## Motivation
- issue number : #9 

## Changes
- DropDown 구현
- BaseVC의 hideKeyboard(=> gesturerecognizer) 가 다른 touch를 막지 못하도록 수정
- CALayer+Extension에 피그마 숫자 값으로 그림자 구현하는 메서드 추가

## Screen Shots
### DropDown 
![ezgif com-gif-maker](https://user-images.githubusercontent.com/77603632/209948164-17404ba3-7367-4165-b0f1-fcedb15b7fea.gif)


####  테스트 코드
```
import UIKit
import FindTownUI

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()

        view.backgroundColor = .white
        let dataSource = ["1", "2","3","4","test5","Test6","test7","test8", "Test9","test10","Test11","test12","Test13","test14"]
        
        let dropDown = DropDown(data: dataSource)
        dropDown.translatesAutoresizingMaskIntoConstraints = false
        
        view.addSubview(dropDown)
        
        NSLayoutConstraint.activate([
            dropDown.centerXAnchor.constraint(equalTo: view.centerXAnchor),
            dropDown.centerYAnchor.constraint(equalTo: view.centerYAnchor),
            dropDown.widthAnchor.constraint(equalToConstant: 200),
            dropDown.heightAnchor.constraint(equalToConstant: 250)
        ])
    }
}
```
## To Reviewers
- tableView는 상위 뷰 보다 크기가 커지면 안되기 때문에 dropdown의 세로 길이가 테이블뷰의 세로 길이가 됩니다!
- scrollIndicator는 색상만 커스텀이 가능한데 그 마저도 지정된 색(검정, 흰색, 회색)만 가능해서 이 부분은 피그마와 동일하게 커스텀이 어려울 것 같습니다 🥹
- DropDown의 항목을 선택했는지 안했는지가 필요한 경우가 생길 수도 있을 것 같아서 (예를 들면 선택해야 버튼이 활성화 된다던가?..) status를 public 으로 설정해두었는데 어떤지 궁금합니당 지금 나온 화면 중에서는 아직 필요한 화면은 없어서요!
- 최대한 비슷하게 하려고 했는데ㅠㅠ 혹시 동작이 어색하거나 버그가 있다면 알려주세용..🥹
- popUp의 버튼 스타일을 수정했는데.. 작은 부분이라 dropdown 브랜치에서 수정했습니다..!😅

seonyoung42 committed:
[fix] tapGesture가 다른 touch를 막지 않도록 수정 #9

seonyoung42 committed:
[feat] DropDown 구현 #9


- 피그마의 숫자 값으로 그림자 표현하는 메서드 CALayer+Extension에 추가
-  dropDown에서 사용되는 이미지 에셋 추가

seonyoung42 committed:
[chore] popUp의 버튼 스타일 변경

seonyoung42 committed:
[fix] dropDown status public으로 변경 #9

ungchun commented on FindTown/FindTownUI/Sources/FindTownUI/DropDown/DropDown.swift:
```diff
@@ -0,0 +1,224 @@
+/// 커스텀 드롭다운, 드롭다운의 세로 길이가 데이터 목록을 보여주는 tableView의 세로 길이가 됨
+public class DropDown: UIView {
+    
+    /// 0번째에는 데이터가 아닌 안내 텍스트 (ex:자치구(선택)) 가 들어감
+    var dataSource = ["test1", "Test2","test3","Test4","test5"]
+    
```
```suggestion
    /// 0번째에는 데이터가 아닌 안내 텍스트 (ex:자치구(선택)) 가 들어감
    var dataSource: [String]
    
```

dataSource를 받으니까 init할 때 받으니까 이 부분은 이렇게 해도 괜찮지 않을까요..?!
아마 지우시는걸 깜빡하신거 같아요!! 😀

seonyoung42 replied:
앗..!!! ㅋㅋㅋㅋ 수정하겠습니다 감사해요 😅

ungchun commented on FindTown/FindTownUI/Sources/FindTownUI/DropDown/DropDownCell.swift:
```diff
@@ -0,0 +1,70 @@
+            self.addSubview($0)
+        }
+
+        NSLayoutConstraint.activate([
+            checkMarkImageview.centerYAnchor.constraint(equalTo: self.centerYAnchor),
+            checkMarkImageview.leadingAnchor.constraint(equalTo: self.leadingAnchor, constant: 11.19),
+            checkMarkImageview.widthAnchor.constraint(equalToConstant: 13.96),
+            checkMarkImageview.heightAnchor.constraint(equalToConstant: 10.48)
+        ])
```
처음에 값들 보고 흠칫했는데 피그마에 저렇게 찍혀있네요 ㅋㅋㅋㅋㅋ 😀😀

ungchun commented on FindTown/FindTownUI/Sources/FindTownUI/Extensions/CALayer+Extension.swift:
```diff
@@ -0,0 +1,34 @@
+import Foundation
+import UIKit
+
+extension CALayer {
+
+    // 피그마값으로 그림자 커스텀할 때 사용하는 함수
+    func customShadow(shadowX: CGFloat,
```
오호.. 그림자 값들을 이렇게 처리하니 좋네요..!! 👍

seonyoung42 replied:
다시보니 함수 네이밍과 파라미터 타입이 불편할 것 같아 조금 수정했어요.. !! 😅

ungchun commented on FindTown/FindTownUI/Sources/FindTownUI/DropDown/DropDown.swift:
```diff
@@ -0,0 +1,224 @@
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    public func addData(data: [String]) {
+        self.dataSource = data
+    }
```
addData 메서드는 init으로 dataSource를 세팅하지만 나중에 addData를 통해서 dataSource를 수정을 할 수 있는 역할을 하는 친구인가요..?!

seonyoung42 replied:
아!! init으로 할 지 addData로 할 지 고민했던.. 흔적입니다 😂 😂 
이번에 왜 이렇게 안 지운 코드가 많은지 모르겠네요 수정하겠습니다 감사해요 !!

ungchun reviewed (commented):
선영님!! 저는 평소에 드롭박스 꽤 어렵다 생각한 부분인데.. 너무 고생하셨습니다..!! 피그마랑 거의 똑같네요..

- status public으로 가도 큰 문제는 없어보여서 괜찮을 거 같아요!!
- 시뮬 돌려봤을 때 버그나 에러는 없었습니다!

seonyoung42 committed:
[fix] DropDown 코드 수정 #9

- dataSource 수정
- addDataSource 메서드 제거
- CALayer+Extension의 addCustomShadow 메서드 수정

seonyoung42 committed:
[fix] DropDownStatus 주석 추가 #9

ungchun reviewed (approved):
고생하셨습니다! 😀

llghdud921 reviewed (changes requested):
선영님!
너무 고생많으셨습니다ㅠㅠ 저도 성훈님과 마찬가지로 dropdown은 정말 까다로운 컴포넌트라고 생각했었어요 정말 대단하세용👍 늦게 리뷰드리는것도 정말 죄송해요ㅠ

다른건 정말 좋은데! 애니메이션이 있으면 정말 좋을 것 같다는 생각이 들어서요!
```swift
UIView.animate(withDuration: 0.3, delay: 0, options: [.allowUserInteraction, .curveEaseOut],
            animations: {
                self.shadowView.alpha = 1
                self.layoutIfNeeded()
        },
        completion: nil)
```
요렇게 alpha 값을 조정하면 animation 효과를 줄수있답니다 ☺️
한번 검토 부탁드릴게요!

seonyoung42 committed:
[fix] shadowView 숨김처리 수정 #9

- isHidden 대신 alpha값을 조정하는 애니메이션으로 수정

llghdud921 reviewed (approved):
선영님! 너무 고생많으셨습니다 👍
